oak_runtime/src/tls.rs

```diff
@@ -14,20 +14,54 @@
 // limitations under the License.
 //
 
+use rustls::{
+    internal::pemfile::{certs, rsa_private_keys},
+    NoClientAuth, PrivateKey, ServerConfig,
+};
 use std::{
     fs::File,
     io::{self, BufReader},
     sync::Arc,
 };
-use tokio_rustls::rustls::{
-    internal::pemfile::{certs, rsa_private_keys},
-    Certificate, NoClientAuth, PrivateKey, ServerConfig,
-};
+
+/// Represents a PEM formatted certificate.
+#[derive(Clone)]
+pub struct Certificate {
+    pem: Vec<u8>,
+}
+
+impl Certificate {
+    /// Checks that the input byte vector represents a valid PEM formatted certificate. If so,
+    /// creates and returns a certificate from the bytes. Otherwise, returns an error.
+    pub fn parse(bytes: Vec<u8>) -> anyhow::Result<Certificate> {
+        let mut cursor = std::io::Cursor::new(bytes.clone());
+
+        // `rustls` doesn't specify certificate parsing errors:
+        // https://docs.rs/rustls/0.17.0/rustls/internal/pemfile/fn.certs.html
+        rustls::internal::pemfile::certs(&mut cursor).map_err(|()| {
+            anyhow::Error::msg("The certificate is not a valid PEM formatted certificate")
+        })?;
+
+        Ok(Certificate { pem: bytes })
+    }
+}
+
+impl AsRef<[u8]> for Certificate {
+    fn as_ref(&self) -> &[u8] {
+        self.pem.as_ref()
+    }
+}
+
+impl From<Certificate> for tonic::transport::Certificate {
+    fn from(certificate: Certificate) -> Self {
+        tonic::transport::Certificate::from_pem(certificate)
+    }
+}
 
 /// Represents TLS identity to use for HTTP server pseudo-nodes.
 #[derive(Default, Clone)]
 pub struct TlsConfig {
-    certs: Vec<Certificate>,
+    certs: Vec<rustls::Certificate>,
```

Comment:
Could / should this be the newly created Certificate type too?

Reply:
I think it is better to keep the `rustls` version here. The other field in `TlsConfig` is `rustls::PrivateKey`, and these fields are used to create other `rustls` objects. I have qualified `PrivateKey` with `rustls` too, for better readability.